### ISSUE TITLE
[RFC] Introduce a Plugin-Mechanism for hardware-specific optimazations

### ISSUE THREAD

**Issue by hahazhky** [pytorch/torchtitan#2183]
**Summary**
This RFC proposes a modular plugin architecture for torchtitan. Inspired by successful extension patterns in vLLM, this mechanism aims to decouple core training logic from hardware-specific optimizations and custom feature extensions (e.g., fused kernel, specialized optimizers, quantization). Currently torchtitan provides some experimental extension points (e.g., in the experimental/ directory or through certain hook interfaces). However, these are insufficient for deep hardware integration

**Proposed Changes**
Firstly, we want to add Kernel-Plugin to enable the integration of platform-specific fused kernel. Including 
1、Dynamic Loading via Entry Points: Utilize Python's entry_points (similar to vLLM) to allow external packages to register themselves automatically.
2、Define the base class CustomOp: Support dispatching and registration (similar to vLLM) 
```
class CustomOp(nn.Module):
    """
    Base class for custom ops.
    Dispatches the forward method to the appropriate backend.
    """

    def __new__(cls, *args, **kwargs):
        try:
            op_name = cls.__name__
        except AttributeError:
            raise TypeError(
                f"Cannot instantiate '{cls.__name__}': its 'name' attribute "
                f"was not set, possibly because it was not decorated with "
                f"@CustomOp.register, or it's the CustomOp base class itself."
            ) from None

        if op_name not in cls.op_registry_oot:
            op_cls_to_instantiate = cls
        else:
            op_cls_to_instantiate = cls.op_registry_oot[op_name]
            logger.debug(
                "Instantiating custom op: %s using %s",
                op_name,
                str(op_cls_to_instantiate),
            )
        return super().__new__(op_cls_to_instantiate)
     ... 
    # Decorator to register custom ops.
    @classmethod
    def register(cls, name: str):
        def decorator(op_cls):
            assert name not in cls.op_registry, f"Duplicate op name: {name}"
            op_cls.name = name
            cls.op_registry[name] = op_cls
            return op_cls

        return decorator
```
3、Modify the model defination using CustomOp, but we can refine this incrementally through subsequent PRs. Alternatively, to minimize the impact on torchtitan's core design, we can just remain current implementation and dynamiclly override by monkey-patch

Feedback is highly welcomed. I’m keen to refine this proposal based on your comments as soon as possible, ensuring we find the best path for hardware extensibility in torchtitan

**Comment by tianyu-l** [comment]
Thanks for the RFC! This sounds an interesting proposal. Let me try to understand the new capabilities:

> Dynamic Loading via Entry Points: Utilize Python's entry_points (similar to vLLM) to allow external packages to register themselves automatically.

Can this be done via `custom_import`? https://github.com/pytorch/torchtitan/blob/main/torchtitan/config/job_config.py#L906

> Define the base class CustomOp: Support dispatching and registration (similar to vLLM)

May I ask what are the ops you'd like to replace?
- For pytorch ops you could override using `torch.library` mechanisms, e.g. see https://docs.pytorch.org/tutorials/advanced/custom_ops_landing_page.html
- For modules you could replace using `ModelConverter` https://github.com/pytorch/torchtitan/blob/main/torchtitan/protocols/model_converter.py#L15

In you proposal, do we need to write every single op/module using `CustomOp`?

> Alternatively, to minimize the impact on torchtitan's core design, we can just remain current implementation and dynamiclly override by monkey-patch

Just to be clear, is there anything torchtitan needs to change to unblock monkey-patching today?

**Comment by hahazhky** [comment]
Thanks for the feedback. 
Q1: yes, I think `custom_import` can handle most of the workload. The main advantage of `Entry Points` is that it removes the need of manual configuration and enable the model to automatically execute on the target hardware.
Q2: I want to replace modules. ModelConverter do works, but this would require modification not only the module itself but also all of its call sites if argument signatures diverge across different hardware backends. For example, if we want to replace RMSNorm/rotary_emb, besides patching kernel interface, we also need to patch the code in class Attention and class TransformerBlock where initialize and call the kernels. **In contrast, CustomOp allows us to enforce a unified interface that guarantees basic compliance across all platforms. We can then leverage `**kwargs` to reserve flexibility for platform-specific extension parameters without breaking the standard API.** 
Q3: we can just modify modules which are often different across hardware-platform, like `MoE`/`RmsNorm`/`Rope`
Q4: no changes are needed. Monkey-patching is **very easy to use** in torchtitan

Thank you again for the review. I look forward to continuing the discussion.

**Comment by tianyu-l** [comment]
@hahazhky

I do have a high-level questions: Is this proposal aiming to support in-torchtitan-repo flexible platform/hardware support? Or out-of-repo (e.g. in your own fork)?

The reason I ask is that if it's in repo, then we'll probably need a way to register multiple implementations + a way to identify hardward/platform and dispatch.
If it is outside the repo, I would assume you only care about another single implementation, which I feel registration is not absolutely necessary and you may just seeking for a more powerful config system.

Right now torchtitan model doesn't support flexible configuring (we support job config, but ModelArgs is kinda rigid). We are thinking of migrating to something like [axlearn](https://github.com/apple/axlearn/blob/main/axlearn/common/config.py#L18) / [olmo](https://github.com/allenai/OLMo/blob/main/olmo/config.py#L235). So if you are only looking at a particular alternative implementation, you could just overwrite the python object (as a config).

It'll make such dependency injection in Q1 less necessary? My worry is that in general registration will make code less readable.
> The main advantage of Entry Points is that it removes the need of manual configuration and enable the model to automatically execute on the target hardware.

I believe such a config system would also solve your Q2/3/4.

**Comment by hahazhky** [comment]
@tianyu-l Thanks for reply.
our primary goal is indeed to support out-of-repo extension, we want to enable hardware-specific optimization without requiring torchtitan main repo changes for every new implementation.
But with `ModelConverter`, I need to do many monkey-patch replacements which seem inelegant and britte, for example, if I want to replace GroupedExperts with optimization version, I need to do conversion like below
```python

class OptimizedGroupedExperts(GroupedExperts):
    pass

class MoEModelConverter(ModelConverter):
    def __init__(self, job_config: JobConfig, parallel_dims: ParallelDims):
        pass

    def convert(self, model: nn.Module):
        """Inplace conversion of the model."""
        for layer in model.layers:
            if layer.moe_enabled:
                # I do some optization with the modification of __init__/forward/init_weights func, 
                # so I need to redefine moe.experts in every TransformerBlock to do the reinitialization
                # Issue: This requires deep knowledge of the TransformerBlock's internal 
                # structure (e.g., '.moe', '.experts', '.moe_enabled'). Any upstream naming change or 
                # structural refactoring will immediately break this converter.
                layer.moe.experts = OptimizedGroupedExperts(layer.moe.experts) 
```
but If we use CustomOp mechanism, I can just focus on the class OptimizedGroupedExperts and GroupedExperts

> Right now torchtitan model doesn't support flexible configuring (we support job config, but ModelArgs is kinda rigid). We are thinking of migrating to something like [axlearn](https://github.com/apple/axlearn/blob/main/axlearn/common/config.py#L18) / [olmo](https://github.com/allenai/OLMo/blob/main/olmo/config.py#L235). So if you are only looking at a particular alternative implementation, you could just overwrite the python object (as a config).

You are right that axlearn/olmo configs solve the parameter flexibility issue. However, the registration mechanism is meant to solve the decoupling and automatic activation of platform-specific optimizations.

**Comment by tianyu-l** [comment]
Thanks for the discussion.

Right now we are still largely in the pattern of "each model defines its own model", except for inner attention kernel and MoE module. The surface sounds too large and unstable to maintain general dependency injection, especially for external use only.

Actually, I'm thinking of consolidating models to share common modules. Maybe after that we can revisit this topic on some of the core modules, e.g. Attention / MoE / RMSNorm / etc.

**Comment by hahazhky** [comment]
OK, thanks for the reply. we will do the extension in our out-of-tree repo currently, we can discuss this later

**Comment by tianyu-l** [comment]
We are thinking more seriously about this recently. Here's a proposal from @syed-ahmed https://gist.github.com/syed-ahmed/169d3bf96eedefd51e4a001a02a617b8

**Comment by syed-ahmed** [comment]
@tianyu-l @felipemello1 I generated the proposal using claude to get a feel for what an end solution might look like. The motivation of the proposal definitely sounds very appealing. We will vet the technical parts a bit more (@tbqh is looking at how https://github.com/pytorch/torchtitan/pull/3266 changes in the context of this plugin system). In the meanwhile, other colleagues have shared the following pointers:
- HF's transformer definition has some interesting extension points for users to customize with their own kernels: https://huggingface.co/docs/kernels/integrating-kernels#transformers
- pytorch/pytorch has recently started integrating vendor kernels: https://github.com/pytorch/pytorch/tree/main/torch/_vendor (and _native kernels although I'm not sure about the specific differences: https://github.com/pytorch/pytorch/tree/main/torch/_native)
- https://github.com/Lightning-AI/lightning-thunder had some original motivations on being able to customize a model without altering their definition written by a model writer.

**Comment by syed-ahmed** [comment]
Capturing some comments from @felipemello1 that didn't get posted in the gist:

> I didn't read it in full, but a couple of thoughts:
> 1. You might want to open this as a PR in titan, so it is easier for people to comment on specific lines
> 2. I wonder how hard it will be to have everything work together, e.g. TorchAO + DeepEP, but composability is usually what many people appreciate
> 3. In Torchtune, we adopted recipes as a way to keep things sane. For example, we have a train_distributed recipe and a train_distributed_QAT recipe for quantization.
> This was great as a way to not mix things up too much in the trainer.py, but the burden of maintenance was much higher, because code+tests were duplicated.

**Comment by hahazhky** [comment]
@tianyu-l Great to see the discussion on the registration mechanism starting. We’ve been following this progress closely and are currently reviewing the new approach mentioned in PR#3396. Looking forward to exploring its feasibility
